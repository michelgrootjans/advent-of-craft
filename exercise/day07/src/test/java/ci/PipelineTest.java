package ci;

import ci.dependencies.Project;
import ci.dependencies.TestStatus;
import org.approvaltests.combinations.CombinationApprovals;
import org.approvaltests.reporters.ClipboardReporter;
import org.approvaltests.reporters.Junit5Reporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.jupiter.api.Test;

import java.util.List;

@UseReporter({ClipboardReporter.class, Junit5Reporter.class})
class PipelineTest {
    @Test
    void run_approval_tests() {
        Boolean[] sendMail = {true, false};
        TestStatus[] testStatuses = {TestStatus.PASSING_TESTS, TestStatus.FAILING_TESTS, TestStatus.NO_TESTS};
        Boolean[] deploySuccedful = {true, false};

        CombinationApprovals.verifyAllCombinations(
            this::run,
            sendMail,
            testStatuses,
            deploySuccedful
        );
    }

    private RunResult run(Boolean sendMail, TestStatus testStatus, boolean deploySuccesful) {
        CapturingLogger logger = new CapturingLogger();
        CapturingMailer mailer = new CapturingMailer();
        Pipeline pipeline = new Pipeline(new TestConfiguration(sendMail), mailer, logger);
        pipeline.run(
            Project.builder()
            .setTestStatus(testStatus)
            .setDeploysSuccessfully(deploySuccesful)
            .build()
        );
        return new RunResult(logger.getLoggedLines(), mailer.sentMails());
    }

    private record RunResult(List<String> loggedLines, List<String> sentMails) {
    }

}
