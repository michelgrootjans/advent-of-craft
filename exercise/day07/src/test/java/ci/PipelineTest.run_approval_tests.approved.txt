[true, PASSING_TESTS, true] => RunResult[loggedLines=[INFO: Tests passed, INFO: Deployment successful, INFO: Sending email], sentMails=[Deployment completed successfully]] 
[true, PASSING_TESTS, false] => RunResult[loggedLines=[INFO: Tests passed, ERROR: Deployment failed, INFO: Sending email], sentMails=[Deployment failed]] 
[true, FAILING_TESTS, true] => RunResult[loggedLines=[ERROR: Tests failed, INFO: Sending email], sentMails=[Tests failed]] 
[true, FAILING_TESTS, false] => RunResult[loggedLines=[ERROR: Tests failed, INFO: Sending email], sentMails=[Tests failed]] 
[true, NO_TESTS, true] => RunResult[loggedLines=[INFO: No tests, INFO: Deployment successful, INFO: Sending email], sentMails=[Deployment completed successfully]] 
[true, NO_TESTS, false] => RunResult[loggedLines=[INFO: No tests, ERROR: Deployment failed, INFO: Sending email], sentMails=[Deployment failed]] 
[false, PASSING_TESTS, true] => RunResult[loggedLines=[INFO: Tests passed, INFO: Deployment successful, INFO: Sending email], sentMails=[Deployment completed successfully]] 
[false, PASSING_TESTS, false] => RunResult[loggedLines=[INFO: Tests passed, ERROR: Deployment failed, INFO: Sending email], sentMails=[Deployment failed]] 
[false, FAILING_TESTS, true] => RunResult[loggedLines=[ERROR: Tests failed, INFO: Sending email], sentMails=[Tests failed]] 
[false, FAILING_TESTS, false] => RunResult[loggedLines=[ERROR: Tests failed, INFO: Sending email], sentMails=[Tests failed]] 
[false, NO_TESTS, true] => RunResult[loggedLines=[INFO: No tests, INFO: Deployment successful, INFO: Sending email], sentMails=[Deployment completed successfully]] 
[false, NO_TESTS, false] => RunResult[loggedLines=[INFO: No tests, ERROR: Deployment failed, INFO: Sending email], sentMails=[Deployment failed]] 
